Handle matrix via reference for grid input

PaulWessel opened the pull request:
The passing of a matrix masquerading as a grid via **GMT_IS_DUPLICATE** [the default mode] now seems to work.  However passing such grids by reference is also allowed, but I suspect there are cases where it cannot be done (since we would need to do something special to the matrix).  Testing is done via testapi_matrix_360_ref.c. Seems to work (please verify if this works from PyGMT, @seisman) but raises an issue related to memory freeing.  Since the matrix data is only pointed to by the grid, I set the grid alloc_mode to external so both of them do not try to free the memory.  However, that also means that the x and y arrays that were created when creating the grid is left at the end.  We could just ignore this rare memory leak or do something about it.  One way is to add info to the struct **GMT_GRID_HEADER_HIDDEN** so that we can free those arrays even those the matrix is left untouched.

I don't know if @joa-quim is wrapping the gmt_hidden.h file in Julia and so adding a new variable will cause trouble.

seisman commented:
I can confirm that https://github.com/GenericMappingTools/pygmt/issues/515 now works for `GMT_IN`, `GMT_IN|GMT_IS_REFERENCE`, and `GMT_IN|GMT_IS_DUPLICATE`. But this PR breaks two PyGMT tests. The two PyGMT tests call grdcut to extract a subregion from a matrix. 

BTW, there is another grdimage bug from PyGMT (https://github.com/GenericMappingTools/pygmt/issues/390). I thought it's caused by the same reason as https://github.com/GenericMappingTools/pygmt/issues/515. But, this PR works for https://github.com/GenericMappingTools/pygmt/issues/515, not for https://github.com/GenericMappingTools/pygmt/issues/390.

PaulWessel commented:
Could you tell me the grdcut arguments? I assume input is a global grid and the subset is not global?  If so then I am not sure why it would break since it would enter the S_obj->region if-test which was there all along.
Less clear why your Py issue 390 would fail after this PR.  it seems exactly like the case herein, but there must be a difference.

seisman commented:
> Could you tell me the grdcut arguments? I assume input is a global grid and the subset is not global? If so then I am not sure why it would break since it would enter the S_obj->region if-test which was there all along.
> Less clear why your Py issue 390 would fail after this PR. it seems exactly like the case herein, but there must be a difference.


Here is the core PyGMT code. It reads the 01d pixel-registered earth relief global grid into an xarray, and extract the subregion of the grid.
```python
grid = load_earth_relief(registration="pixel")
grdcut(grid, outgrid=tmpfile.name, region="0/180/0/90")
```

seisman commented:
> Less clear why your Py issue 390 would fail after this PR. it seems exactly like the case herein, but there must be a difference.

PyGMT issue 390 fails since GMT 6.0, and is worse (crashes or black images) in GMT 6.1, and this PR doesn't fix it.

PaulWessel commented:
I see the report for 390 about zmax < zmin.  This is puzzling since when we just request header info in grdimage, for a matrix, it scans the entire matrix for min/max to be sure it is set.
I don't think we can keep writing C clones to test every possible case in PyGMT.  We need to learn how to run these python scripts and step into the C code in Xcode or whatever debugger.

PaulWessel commented:
And the grdcut case uses reference or duplicate?

seisman commented:
> I see the report for 390 about zmax < zmin. This is puzzling since when we just request header info in grdimage, for a matrix, it scans the entire matrix for min/max to be sure it is set.

It usually means zmin=zmax=0.0.

> I don't think we can keep writing C clones to test every possible case in PyGMT. We need to learn how to run these python scripts and step into the C code in Xcode or whatever debugger.

I already know how to debug PyGMT and C code in Xcode. It's just like what you do with Matlab: run a python console, attach the process id, and run PyGMT codes in the Python console and xcode will stop at the breakpoint. But I don't know where to debug, since the C API is too complicated to understand.

seisman commented:
> And the grdcut case uses reference or duplicate?

Currently, PyGMT always uses `GMT_IN|GMT_IS_REFERENCE`. 

Both `GMT_IN` and `GMT_IN|GMT_IS_DUPLICATE` work, but sometimes they also crash.

PaulWessel commented:
OK, let me give it a try then.  I can check out the pyGMT master on macOS.  Can you give me a quick guide for how to install and select my particular GMT build?

seisman commented:
First of all, you need to install PyGMT dependencies and PyGMT itself. PyGMT will tries to find the system GMT library from the LD_LIBRARY_PATH. 

For debugging purpose, you can just set the environmental variable `GMT_LIBRARY_PATH`, like
```
export GMT_LIBRARY_PATH=~/Gits/gmt/gmt/build/xcode/src/Debug
```
then PyGMT will find the GMT library in this path first.

In the Python console, you can run following commands to verify your GMT library:

```
>>> import pygmt
>>> pygmt.show_versions()
PyGMT information:
  version: v0.1.2+30.gbe75223e
System information:
  python: 3.7.6 (default, Jan  8 2020, 13:42:34)  [Clang 4.0.1 (tags/RELEASE_401/final)]
  executable: /Users/seisman/.anaconda/bin/python
  machine: Darwin-19.6.0-x86_64-i386-64bit
Dependency information:
  numpy: 1.18.1
  pandas: 1.0.1
  xarray: 0.15.1
  netCDF4: 1.5.3
  packaging: 20.1
  ghostscript: 9.50
  gmt: 6.2.0_27ceb6e_2020.08.03
GMT library information:
  binary dir: /Users/seisman/.anaconda/bin
  cores: 8
  grid layout: rows
  library path: /Users/seisman/local/GMT/lib/libgmt.dylib
  padding: 2
  plugin dir: /Users/seisman/local/GMT/lib/gmt/plugins
  share dir: /Users/seisman/local/GMT/share
  version: 6.2.0
```

PaulWessel commented:
I am looking for even more basic info.  I dont see anything on the pyGMT site of installing anything.  I think I need to install conda which I have not used.  Is there an INSTALL page or similar I can just follow?  Sorry to be such a newbie.

seisman commented:
PyGMT has an [installation page](https://www.pygmt.org/latest/install.html). I think you can follow the instruction, but when installing dependencies:
```
conda create --name pygmt python=3.8 pip numpy pandas xarray netcdf4 packaging gmt
```
You can leave the "gmt", because you don't need the gmt package from conda.

PaulWessel commented:
OK, thanks. Do I first install the full Anaconda or is miniconda OK?

seisman commented:
> OK, thanks. Do I first install the full Anaconda or is miniconda OK?

miniconda is OK.

PaulWessel commented:
OK, microconda installed, dependencies installed.  How does python find pygmt on import? Must be some path setting since I get errors if i just run python and enter the import command.

seisman commented:
> OK, microconda installed, dependencies installed. How does python find pygmt on import? Must be some path setting since I get errors if i just run python and enter the import command.

Run
```
cd pygmt
make install
```

PaulWessel commented:
```
(base) pwessel@macnut:~/GMTdev/pygmt-> make install
pip install --no-deps -e .
Obtaining file:///Users/pwessel/GMTdev/pygmt
Installing collected packages: pygmt
  Running setup.py develop for pygmt
Successfully installed pygmt
(base) pwessel@macnut:~/GMTdev/pygmt-> python
Python 3.8.3 (default, May 19 2020, 13:54:14) 
[Clang 10.0.0 ] :: Anaconda, Inc. on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import pygmt
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/pwessel/GMTdev/pygmt/pygmt/__init__.py", line 15, in <module>
    from .session_management import begin as _begin, end as _end
  File "/Users/pwessel/GMTdev/pygmt/pygmt/session_management.py", line 4, in <module>
    from .clib import Session
  File "/Users/pwessel/GMTdev/pygmt/pygmt/clib/__init__.py", line 8, in <module>
    from .session import Session
  File "/Users/pwessel/GMTdev/pygmt/pygmt/clib/session.py", line 10, in <module>
    from packaging.version import Version
ModuleNotFoundError: No module named 'packaging'

```

seisman commented:
> (base) pwessel@macnut:~/GMTdev/pygmt-> make install

It seems you're using the base (the default) conda environment. If you follow the PyGMT install instruction, you should have created a new environment called "pygmt".

Just run `conda activate pygmt` to switch to the newly created enviroment and run `make install` again.

PaulWessel commented:
Sorry, got side-tracked.  OK, verified it found the xcode-built lib.  Trying to run a test by copy paste from test_grdcut.py.  COUld import os, numby but not pytest which is not found.

seisman commented:
Those tests are designed to be run by pytest. You can use this one instead:
```python
import pygmt

from pygmt.datasets import load_earth_relief

grid = load_earth_relief(registration="pixel")
pygmt.grdcut(grid, outgrid="out.grd", region="0/180/0/90")
```

PaulWessel commented:
Thanks for your help, up and running in Xcode now.

joa-quim commented:
> grid = load_earth_relief(registration="pixel")
pygmt.grdcut(grid, outgrid="out.grd", region="0/180/0/90")

Does this mean one have to load the full grid first and only than extract a sub-region?

Tried to reproduce this on Win but it looks my pygmt doesn't work anymore
```
>>> pygmt.show_versions()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: module 'pygmt' has no attribute 'show_versions
```
Note, I refuse to install condas to make py things work (and it worked already)

seisman commented:
> > grid = load_earth_relief(registration="pixel")
> > pygmt.grdcut(grid, outgrid="out.grd", region="0/180/0/90")
> 
> Does this mean one have to load the full grid first and only than extract a sub-region?

For this example, it's yes. But `load_earth_relief` can also accept the *region* (i.e., **-R**) argument, which calls `gmt grdcut @earth_relief_xxy -Rw/e/s/n` and returns a subregion. 

> Tried to reproduce this on Win but it looks my pygmt doesn't work anymore
> 
> ```
> >>> pygmt.show_versions()
> Traceback (most recent call last):
>   File "<stdin>", line 1, in <module>
> AttributeError: module 'pygmt' has no attribute 'show_versions
> ```
> 
> Note, I refuse to install condas to make py things work (and it worked already)

Perhaps you're using an old pygmt? `pygmt.show_versions()` was added in PyGMT 0.1.2, which was released a month ago.

joa-quim commented:
> Perhaps you're using an old pygmt?

Nope, I have only the master source code.

PaulWessel commented:
@seisman I made a few changes in gmt_api.c and the grdcut test writes a correct out.grd.  Perhaps you can check as well.

seisman commented:
> > Perhaps you're using an old pygmt?
> 
> Nope, I have only the master source code.

Please try `pygmt.print_clib_info()` first and see the PyGMT version.

joa-quim commented:
Hmm, nothing about pygmt version
```
>>> pygmt.print_clib_info()
GMT library information:
  binary dir: C:/programs/WPy64-3.7.2/python-3.7.2.amd64
  cores: 12
  grid layout: rows
  library path: C:/programs/GMTdir/gmt_w64.dll
  padding: 2
  plugin dir: C:/programs/GMTdir/gmt_plugins
  share dir: C:/progs_cygw/GMTdev/gmt5/compileds/gmt6/VC14_64/share
  version: 6.2.0
```

seisman commented:
> Hmm, nothing about pygmt version
> 
> ```
> >>> pygmt.print_clib_info()
> GMT library information:
>   binary dir: C:/programs/WPy64-3.7.2/python-3.7.2.amd64
>   cores: 12
>   grid layout: rows
>   library path: C:/programs/GMTdir/gmt_w64.dll
>   padding: 2
>   plugin dir: C:/programs/GMTdir/gmt_plugins
>   share dir: C:/progs_cygw/GMTdev/gmt5/compileds/gmt6/VC14_64/share
>   version: 6.2.0
> ```

Sorry, run `pygmt.__version__` instead.

joa-quim commented:
```
>>> pygmt.__version__
'v0.1.2+30.gbe75223'
```

seisman commented:
> ```
> >>> pygmt.__version__
> 'v0.1.2+30.gbe75223'
> ```

It's weird. You're using the latest version and should be able to run `pygmt.show_versions()`.

seisman commented:
@PaulWessel I think your changes work for `GMT_IN|GMT_IS_REFERENCE`, but for the other two, sometimes the test script crashes. Here are the debug messages:
```
grdcut [DEBUG]: gmt_get_filename: In: 0/180/0/90 Out: 0/180/0/90
grdcut [DEBUG]: Look for file 0/180/0/90 in /Users/seisman/.gmt
grdcut [DEBUG]: Look for file 0/180/0/90 in /Users/seisman/.gmt/cache
grdcut [DEBUG]: Look for file 0/180/0/90 in /Users/seisman/.gmt/server
grdcut [DEBUG]: Got regular w/e/s/n for region (0/180/0/90)
grdcut [INFORMATION]: Processing input grid
grdcut [DEBUG]: gmtapi_begin_io: Input resource access is now enabled [container]
grdcut [DEBUG]: gmtapi_import_grid: Passed ID = 0 and mode = 1
grdcut [DEBUG]: Grid/Image dimensions imply w/e/s/n = -180/180/-90/90, inc = 1/1, pixel registration, n_layers = 1
grdcut [INFORMATION]: Cartesian input grid, yet x spans exactly 360 and -90 <= y <= 90.
grdcut [INFORMATION]:      To make sure the grid is recognized as geographical and global, use the -fg option
grdcut [DEBUG]: Chosen boundary condition for all edges: geographic
grdcut [DEBUG]: Geographic input grid, longitudes span exactly 360
grdcut [DEBUG]: Object ID 1 : Registered Grid Memory Reference 7f8e9fe73610 as an Input resource with geometry Surface [n_objects = 2]
grdcut [DEBUG]: Successfully created a new Grid container
grdcut [DEBUG]: GMT_End_IO: Input resource access is now disabled
grdcut [DEBUG]: gmtapi_begin_io: Input resource access is now enabled [container]
grdcut [DEBUG]: gmtapi_import_grid: Passed ID = 0 and mode = 2
grdcut [INFORMATION]: Importing grid data from user matrix memory location
grdcut [DEBUG]: Grid is considered to have a 360-degree longitude range.
grdcut [DEBUG]: Chosen boundary condition for left and right edges: geographic
grdcut [DEBUG]: Chosen boundary condition for bottom edge: natural
grdcut [DEBUG]: Chosen boundary condition for top    edge: geographic
grdcut [INFORMATION]: Set boundary condition for top    edge: geographic
[1]    4828 segmentation fault  python test.py
```

joa-quim commented:
Ah, it's working now when I launched the python console from within ``C:\progs_cygw\GMTdev\gmt5\pygmt``
Before, I was *somewhere else*

PaulWessel commented:
Sometimes crashes? So if I keep running that grdcut command it will eventually crash? Or do you mean something else?

seisman commented:
> Ah, it's working now when I launched the python console from within `C:\progs_cygw\GMTdev\gmt5\pygmt`
> Before, I was _somewhere else_

I believe it means that you already have pygmt installed, but it's not in the editable/developer mode. So even you have the latest pygmt codes in your pygmt directory, the installed version is the still old one.

To install pygmt in editable mode, you need to cd into the pygmt directory, and run `pip install --no-deps -e .` to install pygmt (here `-e` means editable mode).

seisman commented:
> Sometimes crashes? So if I keep running that grdcut command it will eventually crash? Or do you mean something else?

At line 1274 of `pygmt/clib/session.py`, I changed `GMT_IN|GMT_IS_REFERENCE` to `GMT_IN` or `GMT_IN|GMT_IS_DUPLICATE`, and ran the test script above. Sometimes the script runs fine and the output grid is also correct, but sometimes the same script crashes (see the segment fault error above).

PaulWessel commented:
OK, I changed to duplicate and now no longer the memory leak warnings but no crash after dozens of repeated calls.

PaulWessel commented:
Part of the problems inthis section has to do with GMT doing file-based i/o for decades and then with the API we added other ways.  Initially (pre-mex) we had no way to test any of it.  Mex found tons of basic bugs once we started to pass grids.  Now we are looking more at passing matrix.  We set it up to handle by reference or duplicate.  Duplicate should work since we can copy values into a new grid. However, there are some limitations on how reference can work since we are treating the matrix as readonly.  Let's review those:

1. The matrix is a global grid and we want the whole grid but maybe with a different central meridian.  This was the reason we added testapi_matrix_360.  So this is a possible use of reference since the projection is a separate step.
2. The matrix is used just as is (no region subset).  This should work fine.
3. A subset of the matrix is used.  This also should work fine.  We simply move the boundary inwards by increasing the pad and changing the wesn. This is reversed when done.

However, these two cases cannot work with a readonly matrix:

1. Want a new region that partly overlaps the matrix and party is outside.  We have no way of moving the pad outwards.
2. You want a subset from a global grid and it crosses the periodic boundary,  E.g., reliefs are -180/180 but if I want a cut from 170 to 190 there is no way to achieve that with pad changes.

The solution to this I think is to recognize those two last cases as early as possible inside gmtapi_import_grid and change the method to duplicate and throw in a goto label to get there.

Am I leaving out any other scenarios, @joa-quim and @seisman ?

seisman commented:
> OK, I changed to duplicate and now no longer the memory leak warnings but no crash after dozens of repeated calls.

After several successful runs, I finally catch a crash using `GMT_IN`. Here is the screenshot showing where it crashes. Perhaps you can see something?

![image](https://user-images.githubusercontent.com/3974108/89239230-9a77ba00-d5c6-11ea-94e7-f69afc7cf69e.png)

seisman commented:
Here are the values of the related variables when it crashes:

![image](https://user-images.githubusercontent.com/3974108/89239448-3a354800-d5c7-11ea-8400-8523158d39d1.png)

PaulWessel commented:
OK, this is in the boundary settings.  With duplicate, the grid should have a pad I think.  Just confirming we are still on the grdcut call and this happens inside the call to gmt_BC_init at around line 5486?

seisman commented:
> OK, this is in the boundary settings. With duplicate, the grid should have a pad I think. Just confirming we are still on the grdcut call and this happens inside the call to gmt_BC_init at around line 5486?

Yes, still using the grdcut example in https://github.com/GenericMappingTools/gmt/pull/3829#issuecomment-668284539.

But it crashes at line 12306, in the `gmt_grd_BC_set` function.

PaulWessel commented:
The latest commit takes care of the memory leak and testapi_matrix_360_ref runs without leak complaints now.  However, this does not address the grdcut and other issues.

PaulWessel commented:
Patched a bad bug in the **MATRIX/DUPLICATE** case.  The ij used for the output grid were crazily wrong so your crash is probalby a delayed crash due to stepping on some memory elsewhere.  We now start with output row 0 and not whatever the input row is, and we start with output column 0, not what the input column is.  Could you check if this helps before I return to the **GMT_REFERENCE** case, @seisman ?

seisman commented:
It still crashes sometimes.

PaulWessel commented:
I forgot to check this before the latest commit.  It would be good do know if you could try your current version (without the...f6b commit) and run the grdcut with the region="170/190/0/30" and GMT_IS_REFERENCE.  It should not work - may not crash, but look at the grid and see if a grdimage map of it looks good.
The latest commit deals with the cases I discussed earlier.  If we cannot use reference then we switch to duplicate and jump back to the top and try again as duplicate, which will work.

seisman commented:
> I forgot to check this before the latest commit. It would be good do know if you could try your current version (without the...f6b commit) and run the grdcut with the region="170/190/0/30" and GMT_IS_REFERENCE. It should not work - may not crash, but look at the grid and see if a grdimage map of it looks good.

GMT version: 6.2.0_6e69b36_2020.08.03 + region: "170/190/0/30" + `GMT_IN|GMT_IS_REFERENCE`. The information from grdinfo looks good, but the image (using `gmt grdimage out.grd -Baf -Cgeo -png map`) looks incorrect:

![image](https://user-images.githubusercontent.com/3974108/89250936-dc632900-d5e3-11ea-8710-9e8982b97189.png)

> The latest commit deals with the cases I discussed earlier. If we cannot use reference then we switch to duplicate and jump back to the top and try again as duplicate, which will work.

Using the latest commit gives the correct image (compared with `gmt grdimage @earth_relief_01d -R170/190/0/30`:
![image](https://user-images.githubusercontent.com/3974108/89251149-4da2dc00-d5e4-11ea-844b-05ee53749903.png)

PaulWessel commented:
For matrix and duplication, we have this section that is meant to determine the min/max of the matrix:

```
if (! (mode & GMT_DATA_ONLY)) {	/* Must first init header and copy the header information from the matrix header */
	gmtapi_matrixinfo_to_grdheader (GMT, G_obj->header, M_obj);	/* Populate a GRD header structure */
	if (mode & GMT_CONTAINER_ONLY) {	/* Just needed the header */
		/* Must get the full zmin/max range since not provided by the matrix header */
		gmt_M_grd_loop (GMT, G_obj, row, col, ij) {
			ij_orig = GMT_2D_to_index (row, col, M_obj->dim);
			api_get_val (&(M_obj->data), ij_orig, &d);
			if (gmt_M_is_dnan (d))
				HH->has_NaNs = GMT_GRID_HAS_NANS;
			else {
				G_obj->header->z_min = MIN (G_obj->header->z_min, (gmt_grdfloat)d);
				G_obj->header->z_max = MAX (G_obj->header->z_max, (gmt_grdfloat)d);
			}
		}
		break;	/* Done for now */
	}
}
```

But does not look right to me: If I call GMT_Read_Data with mode GMT_CONTAINER_AND_DATA then I want to get the header set, including the min/max, as well as reading the data. I think in that case, the current code would skip the min/ax step since the (mode & GMT_CONTAINER_ONLY) would be false.  THe only time we do not want to find min/max is when we just want the data, which means we already asked for the header and min/max got set.  I think the (mode & GMT_CONTAINER_ONLY) test should only determine if we call break:

```
if (! (mode & GMT_DATA_ONLY)) {	/* Must first init header and copy the header information from the matrix header */
	gmtapi_matrixinfo_to_grdheader (GMT, G_obj->header, M_obj);	/* Populate a GRD header structure */
	/* Must get the full zmin/max range since not provided by the matrix header */
	gmt_M_grd_loop (GMT, G_obj, row, col, ij) {
		ij_orig = GMT_2D_to_index (row, col, M_obj->dim);
		api_get_val (&(M_obj->data), ij_orig, &d);
		if (gmt_M_is_dnan (d))
			HH->has_NaNs = GMT_GRID_HAS_NANS;
		else {
			G_obj->header->z_min = MIN (G_obj->header->z_min, (gmt_grdfloat)d);
			G_obj->header->z_max = MAX (G_obj->header->z_max, (gmt_grdfloat)d);
		}
	}
	if (mode & GMT_CONTAINER_ONLY)	/* Just needed the header */
		break;	/* Done for now */
}

```
We should enter that loop if we either just want the header or if we want both header and data.  Does that make sense to you @seisman ?

PaulWessel commented:
Thanks for checking the 170/190.  Now that other one.  We are getting closer to the end.

PaulWessel commented:
Also the lines

```
		G_obj->header->z_min = +DBL_MAX;
		G_obj->header->z_max = -DBL_MAX;
```
	
needs to be inside that main if-test, otherwise the min/max get reset when we come to get the data...

seisman commented:
I think you're right with https://github.com/GenericMappingTools/gmt/pull/3829#issuecomment-668365091.

PaulWessel commented:
I've made some additional revisions to this part of the code, especially when we grab the entire matrix.  I also applied what we have learned for the matrix duplicate case to the grid duplicate case which I dont think we are using but it was clearly wrong, just like the matrix initially was clearly wrong.  Let me know if there is any changes from this on the python side.  The tests pass on command line.

seisman commented:
I think your changes have fixed most issues with matrix.

Here is one which doesn't make sense to me:
```
import pygmt
from pygmt.datasets import load_earth_relief

grid = load_earth_relief()
fig = pygmt.Figure()
fig.grdimage(grid, region='g', projection='Q0/0/6i', cmap='geo', frame='a90f30g30')
fig.grdimage(grid, region='g', projection='Q180/0/6i', cmap='geo', X='7i', frame='a90f30g30')
fig.savefig("earth.png")
```
The figures are the same, but the longitudes are wrong:
![image](https://user-images.githubusercontent.com/3974108/89256223-9b721100-d5f1-11ea-874c-e5f92cd9d777.png)

PaulWessel commented:
Will have a look.  I just added another update but unlikely to affect the case above.

PaulWessel commented:
And this is with REFERENCE in session,py?

seisman commented:
REFERENCE and DUPLICATE give the same wrong results

PaulWessel commented:
Soemthing else I need to set?

>>> fig.grdimage(grid, region='g', projection='Q0/0/6i', cmap='geo', frame='a90f30g30')
grdimage [ERROR]: File /Users/pwessel/opt/miniconda3/envs/pygmt/share/geo.cpt was not found

seisman commented:
Yes, when debugging, it seems GMT incorrectly determines the GMT_SHAREDIR, so you need to change GMT_SHAREDIR to the correct GMT/share directory.

PaulWessel commented:
The 180 plot does not work because in this case there is no need for a projection but I cannot rotate the grid without projecting.  So need to find a way to set need_to_project to true when the grid is by reference.

PaulWessel commented:
Your example works for now now.   Let me know if this branch can be merged and I will remove the WIP.

seisman commented:
Find one more crash with `GMT_IN|GMT_IS_REFERENCE`, but it works for `GMT_IN` or `GMT_IN|GMT_IS_DUPLICATE`.

The script reads the file `@tut_relief.nc` into xarray and pass it to grdimage:
```python 
import pygmt
import xarray as xr

pygmt.show_versions()

fig = pygmt.Figure()

fname = pygmt.which("@tut_relief.nc", download='a')
dem = xr.open_dataarray(fname)

fig.grdimage(dem, frame='af', projection='M6i', cmap='geo', shading=True, V='d')

fig.savefig("shading.png")
```

PaulWessel commented:
Unable to reproduce any crash - plot looks good.  Intermittent?

seisman commented:
It always gives me a crash. 

Anyway, I think this PR already fixes the PyGMT issues 390 and 515, and this PR is already very large. 

I think we can merge this PR first. Later I can provide a detailed bug report for the crash above.

PaulWessel commented:
Yes, I was going to suggest something like that as well. Better to focus on a specific - this PR solves many things but possibly not all.
